crates/marketplace/src/testing/order_test.rs

```diff
@@ -0,0 +1,527 @@
+};
+
+use crate::builder_state::MessageType;
+
+use std::{collections::HashSet, fmt::Debug};
+
+use async_compatibility_layer::art::async_sleep;
+use async_std::prelude::FutureExt;
+
+use hotshot_example_types::block_types::TestTransaction;
+
+use crate::{builder_state::TransactionSource, testing::TestTypes};
+use crate::{
+    service::handle_received_txns,
+    testing::{calc_proposal_msg, get_req_msg, start_builder_state},
+};
+use hotshot::rand;
+use std::time::Duration;
+
+/// The function checks whether the common part of two transaction vectors have the same order
+fn order_check<T: Eq + Clone + Debug>(
+    transaction_history: Vec<T>,
+    all_transactions: Vec<Vec<T>>,
+) -> bool {
+    let all_transactions_vec = all_transactions.into_iter().flatten().collect::<Vec<_>>();
+    tracing::debug!(
+        "Doing order check, transaction_history = {:?}, all_transactions = {:?}",
+        transaction_history,
+        all_transactions_vec
+    );
+    let common_txs: Vec<_> = transaction_history
+        .iter()
+        .filter(|item| all_transactions_vec.contains(item))
+        .collect();
+
+    let another_common_txs: Vec<_> = all_transactions_vec
+        .iter()
+        .filter(|item| transaction_history.contains(item))
+        .collect();
+
+    common_txs == another_common_txs
+}
+
+/// This test simulates multiple builder states receiving messages from the channels and processing them
+/// and focus specifically on orders.
+/// It's fine that leader doesn't include some of transactions we've given, or interspersed with other transactions,
+/// as long as the order is correct it will be good.
+#[async_std::test]
+async fn test_builder_order() {
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    tracing::info!("Testing the builder core with multiple messages from the channels");
+
+    // Number of views to simulate
+    const NUM_ROUNDS: usize = 10;
+    // Number of transactions to submit per round
+    const NUM_TXNS_PER_ROUND: usize = 5;
+    // Capacity of broadcast channels
+    const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+    // Number of nodes on DA committee
+    const NUM_STORAGE_NODES: usize = 4;
+
+    let (senders, global_state) = start_builder_state(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+    // Transactions to send
+    let all_transactions = (0..NUM_ROUNDS)
+        .map(|round| {
+            (0..NUM_TXNS_PER_ROUND)
+                .map(|tx_num| TestTransaction::new(vec![round as u8, tx_num as u8]))
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    // generate three different random number between (0..NUM_ROUNDS) to do some changes for output transactions
+    let mut unique_rounds = HashSet::new();
+    #[deny(clippy::modulo_one)]
+    while unique_rounds.len() < 3 {
+        let random_round = std::cmp::max(rand::random::<usize>() % (NUM_ROUNDS - 2), 1);
```

Comment:
The comment above this suggest that we're trying to generate different random numbers between `0..NUM_ROUNDS`, but this statement:

```rust
rand::random::<uszie>() % (NUM_ROUNDS - 2)
```

Would seem to output a random number, `n` | `n` >= 0 and `n` < `NUM_ROUNDS - 2`.

combined with the `max` function then the end result would be:

`n` | `n` > 0 and `n` < `NUM_ROUNDS - 2`.  With `NUM_ROUNDS` = 10, this would mean that `n` > 1 and `n` < 8.  Is this intentional?  If so we should update the comment to indicate this intent.

Reply:
Yeah this is intentional so that round 8 and round 9 could be left to test the case "propose in advance" , let me update the comment.

Reply:
Can we also add a comment for the `n > 1` design?

Reply:
Oh actually, I think it's `n` >= 1 in that case.

Reply:
Oops I found that it's fine to be 0. What I want is `(NUM_ROUNDS - 2)` should be something larger than 1. I'll add an `assert` instead.